app/controllers/payments_controller.rb

```diff
@@ -0,0 +1,92 @@
+class PaymentsController < ApplicationController
+  require 'hips'
+
+  protect_from_forgery except: :webhook
+
+  def create
+    payment_type = params[:type]
+    user_id = params[:user_id]
+
+    @payment = Payment.create(payment_type: payment_type,
+                              user_id: user_id,
+                              status: Payment.order_to_payment_status(nil))
+
+    urls = { success: payment_success_url(user_id: user_id, id: @payment.id) }
+    urls[:error]   = payment_error_url(user_id: user_id, id: @payment.id)
+
+    urls[:webhook] = (SHF_WEBHOOK_HOST || root_url) +
+                     payment_webhook_path.sub('/en', '')
+
+    hips_order = HipsService.create_order(@payment.id,
+                                          user_id,
+                                          session.id,
+                                          payment_type,
+                                          urls)
+    @hips_id = hips_order['id']
+    @payment.hips_id = @hips_id
+    @payment.status = Payment.order_to_payment_status(hips_order['status'])
+    @payment.save
```

Comment:
If there is an error with `@payment.save` with the `rescue` block below work? (_exc_ will be nil, so will `exc.cause` blow up?  (Obv. I haven't tried to make this happen.)  Perhaps this is an indicator we need a test to show that the rescue block will work both if there is a HIPs problem and also if there is a SHF (db save) problem?

Reply:
I was not worried about `save` here as we have already successfully created the record and are just updating attributes.   However, there is a chance of triggering validation errors, so I 1) changed to `save!` and 2) added  `ActiveRecord::RecordInvalid` to the rescue clause.

If there is no chained exception `exc.clause` returns nil.

I'll add a test for rescue capturing a `save!` exception.